### PR TITLE
Introduce and implement `--legacy-array-sections`

### LINKED ISSUE #2443: [SciPy] Code Generation error on passing element of `dimension w(*)` to subroutine call
```fortran
subroutine a(w)
implicit double precision (a-h,o-z)
dimension w(*)
call b(w(icon))
return
end

subroutine b(con)
implicit double precision (a-h,o-z)
dimension con(*)
return
end
```

> LFortran

```console
% lfortran a.f90 --implicit-typing --show-llvm
; ModuleID = 'LFortran'
source_filename = "LFortran"

%array = type { double*, i32, %dimension_descriptor*, i1, i32 }
%dimension_descriptor = type { i32, i32, i32 }

define void @a(%array* %w) {
.entry:
  %icon = alloca i32, align 4
  %0 = load i32, i32* %icon, align 4
  %1 = getelementptr %array, %array* %w, i32 0, i32 2
  %2 = load %dimension_descriptor*, %dimension_descriptor** %1, align 8
  %3 = getelementptr inbounds %dimension_descriptor, %dimension_descriptor* %2, i32 0
  %4 = getelementptr %dimension_descriptor, %dimension_descriptor* %3, i32 0, i32 1
  %5 = load i32, i32* %4, align 4
  %6 = sub i32 %0, %5
  %7 = getelementptr %dimension_descriptor, %dimension_descriptor* %3, i32 0, i32 0
  %8 = load i32, i32* %7, align 4
  %9 = mul i32 %8, %6
  %10 = add i32 0, %9
  %11 = getelementptr %array, %array* %w, i32 0, i32 1
  %12 = load i32, i32* %11, align 4
  %13 = add i32 %10, %12
  %14 = getelementptr %array, %array* %w, i32 0, i32 0
  %15 = load double*, double** %14, align 8
  %16 = getelementptr inbounds double, double* %15, i32 %13
  call void @b(double* %16)
  br label %return

unreachable_after_return:                         ; No predecessors!
  br label %return

return:                                           ; preds = %unreachable_after_return, %.entry
  ret void
}

define void @b(%array* %con) {
.entry:
  br label %return

unreachable_after_return:                         ; No predecessors!
  br label %return

return:                                           ; preds = %unreachable_after_return, %.entry
  ret void
}
code generation error: asr_to_llvm: module failed verification. Error:
Call parameter type does not match function signature!
  %16 = getelementptr inbounds double, double* %15, i32 %13
 %array*  call void @b(double* %16)



Note: Please report unclear or confusing messages as bugs at
https://github.com/lfortran/lfortran/issues.
```

### PR DESCRIPTION
**Pull request by Pranavchiku**
Fixes #2443.

### CODE REVIEW

**Comment by Pranavchiku** [comment]
This PR will need to be strengthen by covering more cases, we can do it as we discover errors.

File: src/lfortran/semantics/ast_body_visitor.cpp [hunk 1]
```diff
@@ -2496,6 +2496,70 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
...
+                            array_arg_idx[i] = ASRUtils::expr_type(f->m_args[i]);
+                        }
+                    }
+                    Vec<ASR::call_arg_t> args_with_array_section;
+                    args_with_array_section.reserve(al, args.size());
+                    for (size_t i = 0; i < args.size(); i++) {
+                        // check if i is in array_arg_idx
+                        if (array_arg_idx.find(i) != array_arg_idx.end()) {
+                            ASR::call_arg_t arg = args[i];
+                            ASR::ttype_t* expected_arg_type = array_arg_idx[i];
```
Comment (Pranavchiku) [inline src/lfortran/semantics/ast_body_visitor.cpp:2513]: This will not work in cases like `dimension SIM(N, *)` where dimension contains variable

**Comment by certik** [comment]
I think that this is good.  Let's find a better name than `--legacy-array-syntax`. Something like "array-slice-by-element" or something like that. Let's think of a good name.

**Comment by Pranavchiku** [comment]
> Let's find a better name than --legacy-array-syntax.

Yes, maybe `--array-item-as-array-slice`? I am poor at naming

**Comment by certik** [comment]
How about `--legacy-array-sections`? More ideas suggested at https://chat.openai.com/share/4125d435-8002-4656-9edd-85fd4f745cb2.

**Comment by Pranavchiku** [comment]
Let's go with `--legacy-array-sections`, it makes sense to me.

**Comment by certik** [comment]
One more thing: can you please add a test in `tests/` that ensures you get a nice error message if the `--legacy-array-sections` is not passed?

I would just add one test into `integration_tests` that just requires one extra option `--legacy-array-sections` and works. Then in `tests/` test the same test without any extra options and ensure you get a nice error message.

I think we should test every such non-default option using the same approach: good error message by default, but working with the option. Both paths tested.

**Comment by Pranavchiku** [comment]
Yes, sure. I will do that.

**Comment by Pranavchiku** [comment]
Minimal example for minpack failure

```fortran
subroutine a(w)
double precision :: w(5)
integer :: icon
call b(w(icon))
return
end

subroutine b(con)
double precision :: con(5)
return
end
```

**Comment by Pranavchiku** [comment]
I feel tests shall pass now.

File: tests/errors/legacy_array_sections1.f90 [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+end
+
+program legacy_array_sections_01
+double precision :: w(5)
+w = 1.39D0
+call a(w)
+print *, w
+if (abs(w(5) - (-4.83D0)) > 1d-8) error stop
+if (abs(w(4) - 3.14D0) > 1d-10) error stop
+end program
```
Comment (certik) [inline tests/errors/legacy_array_sections1.f90:23]: Let's remove this file, and instead test the integration_tests file, which is the same file. That way we ensure exactly the same code will give an error message by default, but works with `--legacy-array-sections`.

File: src/bin/lfortran.cpp [hunk 3]
```diff
@@ -1919,7 +1919,7 @@ int main(int argc, char *argv[])
         std::string arg_kernel_f;
         bool print_targets = false;
         bool link_with_gcc = false;
-	bool fixed_form_infer = false;
+	    bool fixed_form_infer = false;
```
Comment (certik) [inline src/bin/lfortran.cpp:1922]: Remove the white space?

File: tests/errors/legacy_array_sections1.f90 (hunk 2 above)
Comment (Pranavchiku) [reply tests/errors/legacy_array_sections1.f90:23]: Aah yes, I will do that.

File: src/bin/lfortran.cpp (hunk 3 above)
Comment (Pranavchiku) [reply src/bin/lfortran.cpp:1922]: Initially it was not indented properly. I guess after rebasing on top of main, extra tab space got added.

**Pull request merged**